SARAALERT-912: Support Null Flavors in Data Elements

timwongj opened the pull request:
# Description
Jira Ticket: [SARAALERT-912](https://tracker.codev.mitre.org/browse/SARAALERT-912)

Note: Most of this work was previously done by @ashankland.

Update supported values for race (add `Unknown`, `Other`, `Refused to Answer`), ethnicity (add `Unknown`, `Refused to Answer`), and language (add `Other`) for import, API, and UI
  - If `Refused to Answer` or `Unknown` is selected (in the enrollment wizard), gray out unselected race options
  - New values should be added at the end of the import template

# (Feature) Demo/Screenshots
![Screen Shot 2021-02-03 at 4 50 07 PM](https://user-images.githubusercontent.com/9956561/106814110-e839b180-663f-11eb-9928-829d969282c2.png)

# Important Changes
`app/controllers/import_controller.rb`
- add new race fields to import

`app/helpers/fhir_helper.rb`
- add new race fields to fhir API

`app/javascript/components/enrollment/steps/Identification.js`
- add new race options in enrollment

`app/javascript/components/patient/Patient.js`
- add new race options to monitoree details page

`app/jobs/purge_job.rb`
- added new race fields to whitelist fields

`app/lib/import_export.rb`
- added new race fields to import/export

`app/models/concerns/race_validator.rb`
- added separate class for validating race

`db/schema.rb`
- added new race fields to patients table

`public/Sara Alert Import Format.xlsx`
- added new fields to formatting guidance

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [x] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] IE11

Please describe the tests needed to verify your changes. Provide instructions for reproducing these tests. List any relevant details for your test configuration.

tstrass commented:
Needs rebase

timwongj commented:
> Needs rebase

done

tstrass reviewed (changes requested):
Still looking into some of the API stuff but getting this out there now

tstrass commented on app/helpers/fhir_helper.rb:
```diff
@@ -161,39 +176,54 @@ def patient_from_fhir(patient, default_jurisdiction_id)
         url: 'ombCategory',
         valueCoding: FHIR::Coding.new(code: '2076-8', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Native Hawaiian or Other Pacific Islander')
       ) : nil,
+      races[:race_other] ? FHIR::Extension.new(
+        url: 'ombCategory',
+        valueCoding: FHIR::Coding.new(code: 'OTH', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Other')
+      ) : nil,
+      races[:race_unknown] ? FHIR::Extension.new(
+        url: 'ombCategory',
+        valueCoding: FHIR::Coding.new(code: 'UNK', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Unknown')
+      ) : nil,
+      races[:race_refused_to_answer] ? FHIR::Extension.new(
+        url: 'ombCategory',
+        valueCoding: FHIR::Coding.new(code: 'ASKU', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Refused to Answer')
+      ) : nil,
```
Ok so actually I have some other thoughts on how this should be represented in FHIR. [Slack thread](https://mitrecorp.slack.com/archives/CCRMP31C7/p1613512721057800) for context.

Here are the issues as I see things, including with the Ethnicity :
- `"OTH"` is not in the [OMB Race Categories value set](https://www.hl7.org/fhir/us/core/ValueSet-omb-race-category.html#expansion), and so we can't use it like this.
- Neither NullFlavor code, `"UNK"` nor `"ASKU"` is in the [OMB Ethnicity Categories value set](https://www.hl7.org/fhir/us/core/ValueSet-omb-ethnicity-category.html#expansion), and so we can't use them like that for ethnicity.
- There is no ideal way to represent "Other" under the US Core Race Extension

So here are my suggestions:
- Rather than using `"UNK"` and `"ASKU"`, use [Data Absent Reason](https://www.hl7.org/fhir/extension-data-absent-reason.html) values `"unknown"` and `"asked-declined"`. [Here](https://mitrecorp.slack.com/archives/CCRMP31C7/p1613514934060600?thread_ts=1613512721.057800&cid=CCRMP31C7) is a helpful message from Reece about how to use Data Absent Reason. This suggestion applies to Race and Ethnicity.
- For "Other", just use an [`extension:text`](https://www.hl7.org/fhir/us/core/StructureDefinition-us-core-ethnicity-definitions.html#Extension.extension:text) rather than `extension:ombCategory`, using `"other"` as the `valueString`. This suggestion only applies to Race.

Let me know your thoughts.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/a64587cc4e781563136ee93f313d4d74d696d7fb#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6

tstrass replied:
> - Rather than using "UNK" and "ASKU", use Data Absent Reason values "unknown" and "asked-declined". Here is a helpful message from Reece about how to use Data Absent Reason. This suggestion applies to Race and Ethnicity.

I think this would be preferable to using `extension:text` for the `unknown` and `refused_to_answer`.

timwongj replied:
updated: https://github.com/SaraAlert/SaraAlert/pull/689/commits/60eaa418e0bc7f7cb9ee8ca0decb8736ed1ef29d#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R210

tstrass reviewed (changes requested):
Unless I'm missing something, I'm also not seeing an implementation of "Other" as an additional option for language.

timwongj commented on app/models/patient.rb:
```diff
@@ -28,10 +28,6 @@ class Patient < ApplicationRecord
     in: VALID_PATIENT_ENUMS[:monitoring_plan],
     message: "is not an acceptable value, acceptable values are: '#{VALID_PATIENT_ENUMS[:monitoring_plan].reject(&:blank?).join("', '")}'"
   }
-  validates :exposure_risk_assessment, inclusion: {
-    in: VALID_PATIENT_ENUMS[:exposure_risk_assessment],
-    message: "is not an acceptable value, acceptable values are: '#{VALID_PATIENT_ENUMS[:exposure_risk_assessment].reject(&:blank?).join("', '")}'"
-  }
```
@ashankland ?

timwongj replied:
I'm guessing it might of been removed as a result of trying to deal with a merge conflict or something, will add this back for now unless @ashankland mentions a reason

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/a64587cc4e781563136ee93f313d4d74d696d7fb#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292R57

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/a64587cc4e781563136ee93f313d4d74d696d7fb?file-filters%5B%5D=.rb#diff-94dcc014f62f0d9b6084ad98f82f74cd40c86b029ef1691914bdbc960c2f42dcR32

timwongj commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -368,31 +422,74 @@ class Identification extends React.Component {
+                    id="race_refused_to_answer"
+                    label="REFUSED TO ANSWER"
+                    checked={!!this.state.current.patient.race_refused_to_answer}
+                    onChange={this.handleRaceChange}
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/a64587cc4e781563136ee93f313d4d74d696d7fb#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292R484

timwongj commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -43,6 +53,50 @@ class Identification extends React.Component {
   };
 
+  // Special function for handling race selection: If an exclusive race value (e.g. "unknown" or "refused to answer")
+  // is selected, unselects all other checkboxes, and if a standard race value (e.g. "white, black or african amrican", "other")
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/a64587cc4e781563136ee93f313d4d74d696d7fb#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292R57

tstrass reviewed (changes requested):
Still seems there's no addition of an "Other" option to language as the Jira ticket calls for. Was that nixed?

timwongj commented:
> Still seems there's no addition of an "Other" option to language as the Jira ticket calls for. Was that nixed?

Just checked with Kelly, this is something that we do want to add so I'll take care of it

kylietmo commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -43,6 +53,50 @@ class Identification extends React.Component {
+          {}
+        );
+      } else {
+        let test_memo = { [event.target.id]: true };
```
Why is this called `test_memo`?

timwongj replied:
no idea, @ashankland ?

timwongj replied:
this implementation was replaced and is no longer relevant: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292L77

kylietmo commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -43,6 +53,50 @@ class Identification extends React.Component {
+        modified_races = _.reduce(
+          _.union(this.exclusive_race_options, this.default_races),
+          (memo, race) => {
+            return _.extend(memo, { [race]: race == event.target.id });
```
A lot of lodash functions are used here (and below) that I think can just be replaced with using the spread operator. It's kind of hard to read as it is now.

Examples for a few of these lines:
Line 70:
```
{...this.exclusive_race_options, ...this.default_races}
```
Line 72:
```
return {...memo, { [race]: race == event.target.id };
```
Line 91:
```
current: { ...current, patient: { ..._.extend(current.patient, modified_races) } }
```

timwongj replied:
this implementation was replaced and is no longer relevant: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292L77

kylietmo commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -368,31 +422,76 @@ class Identification extends React.Component {
+                    id="white"
+                    label="WHITE"
+                    checked={!!this.state.current.patient.white}
+                    disabled={!!this.state.current.patient.race_unknown || !!this.state.current.patient.race_refused_to_answer}
```
I'd suggest extracting:
```
{!!this.state.current.patient.race_unknown || !!this.state.current.patient.race_refused_to_answer}
```
into a variable at the start rather than having the same logic across all other race checkboxes.

OR actually even better - instead of having separate checkbox elements explicitly listed out here, can we just have a loop that loops over all of the possible races and dynamically creates the `Form.Check`? I think that should be pretty straightforward to do and reduce many many lines of code and the possibility for typos/mistakes.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-0b38c12b7f12dd9f316a576d5b5dc6c12f0b8019720598550b239fca8828d292R411

kylietmo commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -368,31 +422,76 @@ class Identification extends React.Component {
+                    id="race_unknown"
+                    label="UNKNOWN"
+                    checked={!!this.state.current.patient.race_unknown}
+                    disabled={!!this.state.current.patient.race_refused_to_answer}
```
Should "Refused to Answer" be available as an option when "Unknown" is selected (and vice versa)? This check makes me thing it would be but it isn't.

timwongj replied:
it currently isn't available as an option after this change: https://github.com/SaraAlert/SaraAlert/pull/689#discussion_r577063171

what is the desired/expected behavior that we want? should I ask PO team?

kylietmo replied:
I'm not sure myself so yeah I think we should verify with the PO team what is expected.

timwongj replied:
thread has been started here: https://mitrecorp.slack.com/archives/C01A6P94R9B/p1614144113035600

kylietmo replied:
Sounds like it should stay as is. Resolving.

kylietmo commented on app/lib/import_export.rb:
```diff
@@ -72,7 +72,8 @@ module ImportExport # rubocop:todo Metrics/ModuleLength
                                 member_of_a_common_exposure_cohort_type exposure_risk_assessment monitoring_plan exposure_notes full_status symptom_onset
                                 case_status lab_1_type lab_1_specimen_collection lab_1_report lab_1_result lab_2_type lab_2_specimen_collection lab_2_report
-                                lab_2_result jurisdiction_path assigned_user gender_identity sexual_orientation].freeze
+                                lab_2_result jurisdiction_path assigned_user gender_identity sexual_orientation race_other race_unknown
```
So unfortunately this isn't sufficient for making sure these new columns are appended to the exported. On [line 103](https://github.com/SaraAlert/SaraAlert/blob/master/app/lib/import_export.rb#L103) we append `extended_isolation` specifically which makes it show up last in the single monitoree download and all monitorees downloads. I'm not clear on/remembering why this was done with extended isolation, but it'll need to be resolved as we informed jurisdictions their scripts can expect these new columns to be appended at the end.
<img width="685" alt="Screen Shot 2021-02-22 at 4 24 50 PM" src="https://user-images.githubusercontent.com/11698457/108772337-e2423c80-752a-11eb-9cfa-c6784acc89fc.png">

timwongj replied:
Would we be able to just add `extended_isolation_date` as a field in `SARA_ALERT_FORMAT_FIELDS` before these new fields? now that I'm thinking more about this, was `gender_identity` and `sexual_orientation` added after `extended_isolation_date` or the other way around? we probably ran into the same problem if the former was the case but either way, I think it would be good to just add `extended_isolation_field` as an official field to SAF instead of just appending it to the `FULL_HISTORY_PATIENTS_FIELDS` because we're definitely going to be continuing to add more fields in the future

kylietmo replied:
> Would we be able to just add extended_isolation_date as a field in `SARA_ALERT_FORMAT_FIELDS` before these new fields?

I don't see why not, but I feel like there must have been a reason it was added the way it is. Are there any cases you can think of where extended isolation shouldn't be included with the SARA_ALERT_FORMAT_FIELDS? We support it on import + export so I can't think of any myself.

>was gender_identity and sexual_orientation added after extended_isolation_date or the other way around?
Added before I believe, so yeah I think this was a problem then too.

timwongj replied:
here's the PR thread about this: https://github.com/SaraAlert/SaraAlert/pull/336#discussion_r485103442

and here was our slack convo from 9/8: https://mitrecorp.slack.com/archives/D01695F4MF1/p1599589661028500

it sounded like we decided to implement it this way at the time because `extended_isolation` isn't a field that is imported, so we didn't want that to be a field in the Sara Alert Format. We can discuss with Kelly again, but I think this is going to be an ongoing issue where we'll either have to just add this field to SAF or continue to move this field to a different column every time we add new fields (which I think is less ideal)

kylietmo replied:
Nice investigation work - totally forgot about that whole conversation. Okay so for now it doesn't affect the import template (which was my original worry), but it does mean that these new race fields won't be the last column in the export specifically. Yeah this isn't great - one option is that `FULL_HISTORY_PATIENTS_FIELDS` and `FULL_HISTORY_PATIENTS_HEADERS` maybe shouldn't depend on `SARA_ALERT_FORMAT_FIELDS`. Thoughts?

timwongj replied:
that's possible too, but I also think it's very likely that we'll continue to add more fields to `SARA_ALERT_FORMAT_FIELDS` in the future

kylietmo replied:
Agreed, and if we separate out `FULL_HISTORY_PATIENTS_FIELDS` as a separate field that's not dependent on `SARA_ALERT_FORMAT_FIELDS` then that wouldn't be a problem I don't think. The only issue with that is you'd have t update both rather than just one when a field is added to the Sara Alert Format

timwongj replied:
Yeah I think that's another viable solution as well. My only concern with that is with users potentially making the assumption that the formats are the same (other than one of them including `id`), leading to them copying/pasting or writing scripts not being aware of this discrepancy. Should we ask Kelly and @holmesie for thoughts?

kylietmo replied:
What I'm suggesting wouldn't be any change for users, would entirely be on the code side. So if the concern is about there being a very slight difference between the Sara Alert Format export fields vs the Full Patient History export fields then yeah that's been a thing for awhile. So if you want to make sure folks are aware of this difference I don't think it hurts.

timwongj replied:
My concern is the slight difference between the formats, I'll start a slack thread in the UI channel to discuss and get Kelly's input

timwongj replied:
added comment regarding our decision to keep export fields as is here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/5f5f17dfb7ad63738aca9d7b88ed076131222a5b

kylietmo commented on app/controllers/import_controller.rb:
```diff
@@ -222,6 +221,22 @@ def validate_bool_field(field, value, row_ind)
+
+      next unless value && row[race_col].to_s.downcase == 'true'
+
+      err_msg = "'#{VALIDATION[field][:label]}' cannot be true if any other race field is true"
```
This doesn't appear to be working. I can import folks who have "Refused to Answer" as well as "Other" and other non-exclusive races set to true with no validation error.

timwongj replied:
this code is no longer relevant as validation is now done in the `RaceValidator`

kylietmo reviewed (changes requested):
Aside from the comments below, there are a few changes missing:
1. New "Other" option for language (as previously pointed out)
2. In the "Guidance" tab of the updated Sara Alert Import Format file, you'll need to update the possible values for "Ethnicity".
3. You'll want to update this line in the demo rake task: https://github.com/SaraAlert/SaraAlert/blob/master/lib/tasks/demo.rake#L237.
4. @tstrass The the FHIR IG will need to be updated separately due to these changes I think, yeah?

kylietmo commented:
I can't seem get these new values for race to work in the API. I am trying to test the validation in the API by creating a record with both an exclusive and non-exclusive race (ex: “Asian” AND “Unknown”) by doing the following:
```
...
        {
            "extension": [
               {
                    "url": "http://hl7.org/fhir/StructureDefinition/data-absent-reason",
                    "valueCode": "unknown"
                },
                {
                    "url": "ombCategory",
                    "valueCoding": {
                        "system": "urn:oid:2.16.840.1.113883.6.238",
                        "code": "2028-9",
                        "display": "Asian"
                    }
                },
                {
                    "url": "text",
                    "valueString": "Unknown, Asian"
                }
            ],
            "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race"
        },
....
```
but I can’t seem to get it to create a Patient with an Unknown value (even if I exclude the other race here). @timwongj Were you able to get this to work locally?

tstrass commented on app/helpers/fhir_helper.rb:
```diff
@@ -161,39 +176,54 @@ def patient_from_fhir(patient, default_jurisdiction_id)
         url: 'ombCategory',
         valueCoding: FHIR::Coding.new(code: '2076-8', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Native Hawaiian or Other Pacific Islander')
       ) : nil,
+      races[:race_other] ? FHIR::Extension.new(
+        url: 'text',
+        valueString: 'other'
+      ) : nil,
```
Realizing my original suggestion to represent "Other" this way won't work, because the `extension:text` within the US Core Race extension is intended to describe all of the Patient's race concept's, and there can only be one (which we have below in lines 219-226). My new suggestion is to use `extention:detailed`, using the "Other Race" code from the [detailed race value set](https://www.hl7.org/fhir/us/core/ValueSet-detailed-race.html). I was trying to avoid this before since that code is deprecated within HL7, but I see no other way.

```suggestion
      races[:race_other] ? FHIR::Extension.new(
        url: 'detailed',
        valueCoding: 'FHIR::Coding.new(code: '2131-1', system: 'urn:oid:2.16.840.1.113883.6.238', display: 'Other Race')'
      ) : nil,
```

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R209

tstrass commented on app/helpers/fhir_helper.rb:
```diff
@@ -134,6 +146,9 @@ def patient_from_fhir(patient, default_jurisdiction_id)
       american_indian_or_alaska_native: race_code?(patient, '1002-5'),
       asian: race_code?(patient, '2028-9'),
       native_hawaiian_or_other_pacific_islander: race_code?(patient, '2076-8'),
+      race_other: race_code?(patient, 'OTH'),
+      race_unknown: race_code?(patient, 'UNK'),
+      race_refused_to_answer: race_code?(patient, 'ASKU'),
```
This needs to be updated per the corresponding changes you made in the `to_us_core_race` method. I believe this explains the issue Kylie noted [here](https://github.com/SaraAlert/SaraAlert/pull/689#issuecomment-783705532).

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R152

tstrass commented on app/helpers/fhir_helper.rb:
```diff
@@ -232,6 +270,8 @@ def from_us_core_ethnicity(patient)
     code = patient&.extension&.select { |e| e.url.include?(url) }&.first&.extension&.select { |e| e.url == 'ombCategory' }&.first&.valueCoding&.code
     return 'Hispanic or Latino' if code == '2135-2'
     return 'Not Hispanic or Latino' if code == '2186-5'
+    return 'Unknown' if code == 'UNK'
+    return 'Refused to Answer' if code == 'ASKU'
```
This also needs to be updated per the corresponding changes you made in the `to_us_core_ethnicity` method.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R277

timwongj commented on app/controllers/import_controller.rb:
```diff
@@ -222,6 +221,22 @@ def validate_bool_field(field, value, row_ind)
     raise ValidationError.new(err_msg, row_ind)
   end
 
+  def validate_exclusive_race_field(field, row, col_num, row_ind)
```
is this method no longer needed if we're calling `RaceValidator` upon import which does its own validation?

kylietmo replied:
Yeah no longer needed, it depended on which PR was merged first. See related comment below: 
> If @ngfreiter's PR for import validation consolidation goes in before this then this should be updated to be used on both API and import, and the new method in the import controller won't be needed. So we'll need to keep an eye out for whichever one goes in first, as there will need to be changes either way.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-94dcc014f62f0d9b6084ad98f82f74cd40c86b029ef1691914bdbc960c2f42dcR105

timwongj commented on app/models/concerns/race_validator.rb:
```diff
@@ -0,0 +1,14 @@
+    has_non_exclusive_race = record.white || record.black_or_african_american || record.american_indian_or_alaska_native ||
+                             record.asian || record.native_hawaiian_or_other_pacific_islander || record.race_other
+
+    return unless has_non_exclusive_race ? (record.race_unknown || record.race_refused_to_answer) : (record.race_unknown && record.race_refused_to_answer)
```
The first case (where `has_non_exclusive_race` is `true`) uses an `||` because race cannot be a non_exclusive_race + either one of the exclusive ones, the second case uses a `&&` because none of the non_exclusive_races are selected, which means it's only invalid if both `unknown` and `refused_to_answer` are selected, but it's valid if only one of them are selected. Does that make sense or am I missing something?

kylietmo replied:
Yep - mistake on my part. Thanks!

timwongj commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -24,6 +25,15 @@ class Identification extends React.Component {
       modified: {},
       languageOptions: this.getLanguageOptions(),
     };
+    this.default_races = [
```
I'm more than happy to do that however, that might lead to these constant arrays having to be defined in 3 different instances for `new`, `edit`, and `new_group_member` whether we define it in the erb files or the controller, would it make sense to define these in `validation_helper` or something once and then pass them as props?

kylietmo replied:
Yeah that's what I'm thinking - ideally have these in one location

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-4f61e388997f8d92e41f66947a47b461707b615cf9fac69f1080b8254630d2ceR92

timwongj commented on app/helpers/validation_helper.rb:
```diff
@@ -149,6 +149,9 @@ def self.normalize_enums(enums_dict)
     native_hawaiian_or_other_pacific_islander: { label: 'Native Hawaiian or Other Pacific Islander', checks: [:bool] },
+    race_other: { label: 'Other', checks: [:bool] },
+    race_unknown: { label: 'Unknown', checks: [:bool] },
+    race_refused_to_answer: { label: 'Other', checks: [:bool] },
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-4f61e388997f8d92e41f66947a47b461707b615cf9fac69f1080b8254630d2ceR171

timwongj commented on app/javascript/components/public_health/PublicHealthHeader.js:
```diff
@@ -127,7 +127,7 @@ class PublicHealthHeader extends React.Component {
             <div className="mb-3">
               <a href="https://github.com/SaraAlert/SaraAlert/blob/master/public/Sara%20Alert%20Import%20Format.xlsx?raw=true">Download formatting guidance</a>{' '}
-              (Updated 2/9/2021)
+              (Updated 2/23/2021)
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-773ded7d556c74e27b06626fcca3ea85b98c4f15930b07e36b4dc923b291841aR130

timwongj commented:
> Still seems there's no addition of an "Other" option to language as the Jira ticket calls for. Was that nixed?

addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-ad6c022514987377834d6d99a324222ae2acb2e75542610af9d5c0abf1c8cbc6R284

timwongj commented:
> Aside from the comments below, there are a few changes missing:
> 
> 1. New "Other" option for language (as previously pointed out)
> 2. In the "Guidance" tab of the updated Sara Alert Import Format file, you'll need to update the possible values for "Ethnicity".
> 3. You'll want to update this line in the demo rake task: https://github.com/SaraAlert/SaraAlert/blob/master/lib/tasks/demo.rake#L237.
> 4. @tstrass The the FHIR IG will need to be updated separately due to these changes I think, yeah?

1. Addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-ad6c022514987377834d6d99a324222ae2acb2e75542610af9d5c0abf1c8cbc6R284
2. Addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#
3. Addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-9bfc38ef71b77112eb91f77e35668d17dfea639c79e19de03900b201ad44b8efR274

timwongj commented on app/models/patient.rb:
```diff
@@ -88,6 +89,8 @@ class Patient < ApplicationRecord
 
   validates_with PrimaryContactValidator, on: :api
 
+  validates_with RaceValidator, on: :api
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-94dcc014f62f0d9b6084ad98f82f74cd40c86b029ef1691914bdbc960c2f42dcR105

timwongj commented on app/models/concerns/race_validator.rb:
```diff
@@ -0,0 +1,14 @@
+# Validates that a given date (attribute) is valid
+class RaceValidator < ActiveModel::Validator
+  def validate(record)
+    # If value is unsuccessfully typecast to a date, it will be nil, so validate on the value before cast
```
addressed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/767b8765f1830cb0e34e8bc8b1adbbe48c0786f5#diff-4004341efe28eb2594fef71f4e7df2724db98e19bf1d781636315b0e50cb8762L6

timwongj commented:
> I can't seem get these new values for race to work in the API. I am trying to test the validation in the API by creating a record with both an exclusive and non-exclusive race (ex: “Asian” AND “Unknown”) by doing the following:
> 
> ```
> ...
>         {
>             "extension": [
>                {
>                     "url": "http://hl7.org/fhir/StructureDefinition/data-absent-reason",
>                     "valueCode": "unknown"
>                 },
>                 {
>                     "url": "ombCategory",
>                     "valueCoding": {
>                         "system": "urn:oid:2.16.840.1.113883.6.238",
>                         "code": "2028-9",
>                         "display": "Asian"
>                     }
>                 },
>                 {
>                     "url": "text",
>                     "valueString": "Unknown, Asian"
>                 }
>             ],
>             "url": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race"
>         },
> ....
> ```
> 
> but I can’t seem to get it to create a Patient with an Unknown value (even if I exclude the other race here). @timwongj Were you able to get this to work locally?

I pushed some changes according to some things @tstrass pointed out and it's hopefully fixed now.

I actually haven't been testing the API stuff (not sure how to set up an environment for that). Should I use postman to try to run it locally? Is there anything I need to do to setup for authentication?

kylietmo commented:
> I pushed some changes according to some things @tstrass pointed out and it's hopefully fixed now.
> 
> I actually haven't been testing the API stuff (not sure how to set up an environment for that). Should I use postman to try to run it locally? Is there anything I need to do to setup for authentication?

Yep POSTMAN works great. You can just test with the user workflow pretty easily by grabbing an auth code [here](https://saraalert.github.io/SaraAlert/api/api-getting-started.html#testing) and setting up a request for an access token as explained [here](https://saraalert.github.io/SaraAlert/api/api-getting-started.html#post-oauthtoken). Then make a POST/PUT/etc as explained [here](https://saraalert.github.io/SaraAlert/api/api-specification.html#post-basepatient). I'd be happy to get in a quick Slack call to show you the setup, as I'm sure Nick or Tom would be as well

kylietmo commented:
> 2\. In the "Guidance" tab of the updated Sara Alert Import Format file, you'll need to update the possible values for "Ethnicity".

@timwongj Could you update the date in the "Updates" tab as well to be 3/9? Also annoying thing about excel, but it opens on the tab and with the selection you had when you saved it, so could you also select the the leftmost cell of each tab and save on the "Updates" tab?

kylietmo commented on app/helpers/fhir_helper.rb:
```diff
@@ -230,10 +268,13 @@ def to_us_core_ethnicity(ethnicity)
     return 'Hispanic or Latino' if code == '2135-2'
     return 'Not Hispanic or Latino' if code == '2186-5'
+    return 'Unknown' if code == 'unknown'
+    return 'Refused to Answer' if code == 'asked-declined'
```
Can't comment on the line directly, but right now the new ethnicity values are not updated in this line for the API: https://github.com/SaraAlert/SaraAlert/blob/07ea552442160d2386e34d5ae0cfcdb676fa38de/app/helpers/fhir_helper.rb#L210

I feel like this line should be pulling from the values in ValidationHelper also so it's one less place we have to update in the future.

timwongj replied:
fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/5bf58817fd305885bab670ab491e78dea615eeaf#diff-402d451d1f7019322100522f6898c3153444edb8ddb963e181a457c84a2ca6b6R242

> > 2. In the "Guidance" tab of the updated Sara Alert Import Format file, you'll need to update the possible values for "Ethnicity".
> 
> @timwongj Could you also update the date in the "Updates" tab as well to be 3/9? Also annoying thing about excel, but it opens on the tab and with the selection you had when you saved it, so could you also select the the leftmost cell of each tab and save on the "Updates" tab?

fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/f87317c97428ade207f5630c951f2c0a9dd827b8

timwongj commented on app/javascript/data/supportedLanguages.json:
```diff
@@ -279,6 +279,14 @@
         "email": false,
         "phone": false
       }
+    },
```
hmm, how should `Other` be handled in FHIR? It's not on this list for languages: http://hl7.org/fhir/stu3/valueset-languages.html

@tstrass any suggestions?

timwongj replied:
resolving this thread since we will not be handling languages in this ticket/PR

timwongj commented on app/javascript/components/enrollment/steps/Identification.js:
```diff
@@ -43,6 +43,45 @@ class Identification extends React.Component {
+      }
+
+      races_to_reset.forEach(option => {
+        modified_races[`${option.race}`] = false;
```
unless I'm missing something, I don't think this line is modifying state? `modified_races` is a separate variable defined on line 51

kylietmo replied:
Whoops, thought this said `modified` rather than `modified_races` - all set!

timwongj commented on app/lib/import_export.rb:
```diff
@@ -96,11 +97,12 @@ module ImportExport # rubocop:todo Metrics/ModuleLength
 
   FULL_HISTORY_PATIENTS_FIELDS = ([:id] + SARA_ALERT_FORMAT_FIELDS + [:extended_isolation]).freeze
 
-  FULL_HISTORY_PATIENTS_HEADERS = (['Patient ID'] + SARA_ALERT_FORMAT_HEADERS + ['Extended Isolation Date']).freeze
```
this was done on accident (started messing moving the extended isolation date to SAF but decided to undo it until we had a verdict), fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/f87317c97428ade207f5630c951f2c0a9dd827b8#diff-c34521fcd0fdc2034d456886aae8fd35299fe908ab88d8e49c60d0b4d39ab26bR105

timwongj commented on app/helpers/validation_helper.rb:
```diff
@@ -89,6 +89,21 @@ module ValidationHelper # rubocop:todo Metrics/ModuleLength
+      { race: :race_refused_to_answer, label: 'REFUSED TO ANSWER' }
+    ]
+  }.freeze
+
```
fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/f87317c97428ade207f5630c951f2c0a9dd827b8#diff-4f61e388997f8d92e41f66947a47b461707b615cf9fac69f1080b8254630d2ceR117

timwongj commented:
> > 2. In the "Guidance" tab of the updated Sara Alert Import Format file, you'll need to update the possible values for "Ethnicity".
> 
> @timwongj Could you also update the date in the "Updates" tab as well to be 3/9? Also annoying thing about excel, but it opens on the tab and with the selection you had when you saved it, so could you also select the the leftmost cell of each tab and save on the "Updates" tab?

fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/f87317c97428ade207f5630c951f2c0a9dd827b8

kylietmo commented on app/javascript/components/enrollment/steps/Contact.js:
```diff
@@ -14,12 +14,12 @@ const phoneUtil = libphonenumber.PhoneNumberUtil.getInstance();
   constructor(props) {
     super(props);
-    this.state = { ...this.props, current: { ...this.props.currentState }, errors: {}, modified: {}, isEditMode: window.location.href.includes('edit') };
+    this.state = { ...this.props, current: { ...this.props.currentState }, errors: {}, modified: {}, isedit_mode: window.location.href.includes('edit') };
```
I think these changes snuck in from elsewhere.

timwongj replied:
yep, it snuck in when I was trying to fix conflicts for a rebase, fixed here: https://github.com/SaraAlert/SaraAlert/pull/689/commits/79f552e73ee0c551a9718f213a2a3d95ded2e0d3

kylietmo commented:
When you rebase and resolve conflicts, please also squash commits.

kylietmo commented on app/javascript/components/public_health/PublicHealthHeader.js:
```diff
@@ -127,7 +127,7 @@ class PublicHealthHeader extends React.Component {
             <div className="mb-3">
               <a href="https://github.com/SaraAlert/SaraAlert/blob/master/public/Sara%20Alert%20Import%20Format.xlsx?raw=true">Download formatting guidance</a>{' '}
-              (Updated 2/9/2021)
+              (Updated 3/9/2021)
```
Needs to be updated to 3/23/2021. Same with the dates in the import template.

timwongj replied:
fixed: https://github.com/SaraAlert/SaraAlert/pull/689/commits/39ae0cc24006c7ef696115836d7421f42e95429c

kylietmo replied:
Looks like you updated the import template (thanks!), however this line will also need to be updated

timwongj replied:
sorry, thought I updated it when I didn't, should be fixed now

kylietmo reviewed (changes requested):
Everything is looking good. Just some changes in the template file are needed and then I should be ready to approve.

In the import template, in the "Updates" tab, it only mentions the changes to the race field but technically the ethnicity field has also been updated. I'd recommend adding something like "Ethnicity field now supports two new values: 'Unknown' and 'Refused to Answer'" and then add that to the list of affected fields. 

Also @tstrass who is updating the FHIR IG with these changes?

timwongj commented on app/helpers/fhir_helper.rb:
```diff
@@ -163,53 +178,68 @@ def patient_from_fhir(patient, default_jurisdiction_id)
-    url = 'us-core-race'
-    patient&.extension&.select { |e| e.url.include?(url) }&.first&.extension&.select { |e| e.url == 'ombCategory' }&.any? { |e| e&.valueCoding&.code == code }
+  def race_code?(patient, code, url, absent)
+    race_extension = patient&.extension&.select { |e| e.url.include?('us-core-race') }&.first&.extension
```
fixed: https://github.com/SaraAlert/SaraAlert/pull/689/commits/39ae0cc24006c7ef696115836d7421f42e95429c

tstrass commented:
Needs rebase

timwongj commented:
> Needs rebase, other than that this LGTM

Done
